### PR TITLE
feature #520: Thread safe User API exposure of ALL Consumer APIs

### PR DESCRIPTION
**Pull request by astubbs**
Fixes:
- #520 

### Checklist

- [ ] Documentation
- [ ] Changelog

Fixes:
- #191 
- #186

Blocked by:
- #270 
- #325

### CODE REVIEW

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ConsumerApiAccess.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ * @author Antony Stubbs
+ * @see FullConsumerFacade
+ * @see PCConsumerAPI
+ * @see ConsumerFacade
+ */
+public interface ConsumerApiAccess<K, V> {
+
+    /**
+     * todo docs
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ConsumerApiAccess.java:19]: docs

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    /**
+     * todo docs
+     */
+    boolean swallow = false;
+
+    public FullConsumerFacade(BrokerPollSystem<K, V> basePollerRef) {
+        super(basePollerRef);
+    }
+
+    // todo branch feature - optionally have the implememtation talk directly to the work manager and "poll" records from it - PC can look exactly like a normal consumer?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java:43]: resolve

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        return ConsumerRecords.empty();
+    }
+
+    @Override
+    public ConsumerRecords<K, V> poll(final Duration timeout) {
+        throwInvalidCall();
+        return ConsumerRecords.empty();
+    }
+
+    // todo any universe where allowing direct offset committing makes sense? might act like a sort of seek?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java:56]: resolve

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    public void commitSync(Map<TopicPartition, OffsetAndMetadata> offsets, Duration timeout) {
+        throwInvalidCall();
+    }
+
+    @Override
+    public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
+        throwInvalidCall();
+    }
+
+    // todo pass through to manually "blocking" or even really pausing partitions?
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java:72]: resolve

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    @Override
+    public void resume(final Collection collection) {
+        throwInvalidCall();
+    }
+
+    // audit
+    @Override
+    public void enforceRebalance() {
+        throwInvalidCall(); // ?
+//        return blockingAskConsumer(consumer -> consumer.enforceRebalance(partition, timeout));
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/FullConsumerFacade.java:87]: resolve - allow or not?

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCConsumerAPI.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+    @SneakyThrows
+    ConsumerGroupMetadata groupMetadata();
+//
+//    /**
+//     * Closes the entire PC system, not just the consumer.
+//     *
+//     * @see ParallelEoSStreamProcessor#close
+//     */
+//    void close();
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCConsumerAPI.java:88]: resolve

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCConsumerAPI.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+ * <p>
+ * Generally, you can only... And you can't...
+ * <p>
+ * All methods are / must be thread safe.
+ *
+ * @param <K>
+ * @param <V>
+ * @author Antony Stubbs
+ */
+public interface PCConsumerAPI {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/internal/PCConsumerAPI.java:32]: javadoc link to Consumer interface for method javadocs

File: parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/internal/ConsumerFacadeTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+import java.util.Set;
+
+import static io.confluent.parallelconsumer.ManagedTruth.assertTruth;
+
+/**
+ * @author Antony Stubbs
+ * @see ConsumerFacade
+ */
+class ConsumerFacadeTest extends BrokerIntegrationTest {
```
Comment (astubbs) [inline parallel-consumer-core/src/test-integration/java/io/confluent/parallelconsumer/internal/ConsumerFacadeTest.java:20]: more testing

File: parallel-consumer-examples/parallel-consumer-example-core/src/main/java/io/confluent/parallelconsumer/examples/core/CoreApp.java [hunk 9]
```diff
@@ -55,6 +56,14 @@ void run() {
 
         postSetup();
 
+        // example consumer API access
```
Comment (astubbs) [inline parallel-consumer-examples/parallel-consumer-example-core/src/main/java/io/confluent/parallelconsumer/examples/core/CoreApp.java:59]: resolve

File: parallel-consumer-examples/pom.xml [hunk 10]
```diff
@@ -24,4 +24,13 @@
         <module>parallel-consumer-example-reactor</module>
     </modules>
 
+    <dependencies>
+        <dependency>
+            <groupId>io.stubbs.truth</groupId>
```
Comment (astubbs) [inline parallel-consumer-examples/pom.xml:29]: needed?

**Review by astubbs — COMMENTED** [review]
looking good!

**Comment by eddyv** [comment]
Closing - Stale.

**Pull request closed without merging**